### PR TITLE
Add `--empty-parent` and `--empty-child` to `jj split`

### PR DESCRIPTION
**Pull request by samueltardieu**
Following the discussion in #1155, I tried an implementation of `--empty-parent` (`-P`) and `--empty-child` (`-C`) with `jj split`. They do not prompt for a description: the non-empty change keeps the original description and change id. The empty change gets an empty description and a new change id.

The fact that the non-empty commit keeps the change id when using `--empty-parent` differs from what is done with `jj split nonexistent`, but it seems more logical to me. I've done this in a separate commit as to be able to revert it easily if needed.

# Checklist

If applicable:
- [X] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [X] I have added tests to cover my changes

### CODE REVIEW

File: src/commands/mod.rs
Comment (ilyagr) [inline src/commands/mod.rs:2643]: In the case of `--empty-parent` (only), this commit should get a new change_id while `second_commit` below shouldn't.

File: tests/test_split_command.rs
Comment (ilyagr) [inline tests/test_split_command.rs:169]: Optional: You could make "target commit" have more than one parent.

File: tests/test_split_command.rs
Comment (ilyagr) [inline tests/test_split_command.rs:198]: Optional: You could do a `jj undo` here and then include the second test, which would be shorter. I also keep finding weird bugs with `jj undo` by structuring tests this way. I find this doesn't make tests less readable. Because of `insta`, this usually doesn't cause the entire test to fail when a piece of it fails (though sometimes it does happen). If the test gets long, I just insert another `jj log` after the `jj undo`.

**Review by ilyagr — CHANGES_REQUESTED** [review]
This approach (adding options to `jj split`) looks fine to me, but we should wait for @martinvonz and/or @yuja to comment.

I think there's an annoying change needed to which commit gets which change id. There's no reason to fix that unless we officially settle on this approach. Other than that, this looks good to me; I just have some minor take-it-or-leave-it comments.

File: src/commands/mod.rs
Comment (ilyagr) [reply src/commands/mod.rs:2643]: I realized this also raises the question of whether branches should go with the child even on `--empty-parent`. It's a bit arguable, but I think for `jj split` the answer should be "yes". So, just changing where the id goes and not worrying about the branches would do the right thing.

**Comment by martinvonz** [comment]
My main concern with `--empty-parent` and `--empty-child` is that `split` is probably not the command I would reach for to insert a new commit in a stack. OTOH, if the user actually did look at `jj help split` and wanted to create an empty child or parent, it does seem better to have flags for that than to have them figure out the `jj split .` and `jj split nonexistent` tricks. So I'm not sure what's best. I also left [a comment](https://github.com/martinvonz/jj/issues/1155#issuecomment-1411492637) on #1155.

File: src/commands/mod.rs
Comment (samueltardieu) [reply src/commands/mod.rs:2643]: > In the case of `--empty-parent` (only), this commit should get a new change_id while `second_commit` below shouldn't.

This is done in the second commit of this issue.

File: lib/src/commit_builder.rs
Comment (ilyagr) [inline lib/src/commit_builder.rs:122]: I don't think it's worth modifying the CommitBuilder class for this.

We could do it the boring and somewhat ugly way without an `apply` function. Alternatively, we could get some inspiration from https://docs.rs/apply/latest/src/apply/lib.rs.html#16-31. I don't know it's better to depend on that crate or to include a similar function inside our library.

File: src/commands/mod.rs
Comment (ilyagr) [reply src/commands/mod.rs:2643]: Thank you!

**Comment by yuja** [comment]
> My main concern with `--empty-parent` and `--empty-child` is that `split` is probably not the command I would reach for to insert a new commit in a stack.

I feel the same. Functionality wise, these `jj split` options work for me. But maybe successors/predecessors chain would be different from what I expect for "new" commit insertion?

File: CHANGELOG.md
Comment (samueltardieu) [reply CHANGELOG.md]: Done.

File: src/commands/mod.rs
Comment (samueltardieu) [reply src/commands/mod.rs]: Done.

File: src/commands/mod.rs
Comment (samueltardieu) [reply src/commands/mod.rs]: Done.

File: src/commands/mod.rs
Comment (samueltardieu) [reply src/commands/mod.rs:2595]: Indeed, since `String::new()` does not allocate it is appropriate. Done.

File: tests/test_split_command.rs
Comment (samueltardieu) [reply tests/test_split_command.rs]: Done with a `get_log_output()` helper.

File: tests/test_split_command.rs
Comment (samueltardieu) [reply tests/test_split_command.rs]: Good idea. Done.

File: tests/test_split_command.rs
Comment (samueltardieu) [reply tests/test_split_command.rs:172]: I have added a helper to create the initial setup (two parents, target commit, two children) so the code is lighter. But I'll keep the `undo` in mind.

File: lib/src/commit_builder.rs
Comment (samueltardieu) [reply lib/src/commit_builder.rs]: Done.

File: CHANGELOG.md
Comment (ilyagr) [inline CHANGELOG.md:108]: This needs a comma and has too many "after"s. "an empty parent before, or an empty child after..."

**Review by ilyagr — APPROVED** [review]
I'm approving this to get rid of the annoying red "changes requested" marker. Please don't merge it without an OK from others.

**Comment by samueltardieu** [comment]
This PR is suspended at the moment as it is less powerful than #1172 although less intimidating. Let's wait until #1172 is settled.

File: CHANGELOG.md
Comment (samueltardieu) [reply CHANGELOG.md]: Done

**Pull request closed without merging**
